[Lens] Track actions in the UI by time

wylieconlon opened the pull request:
This PR introduces:

* A telemetry collector that runs on a 10 second interval while the Lens app is open. Events are tracked to local storage, and if there are any events captured in 10 seconds, they are posted to the server in a batch. Events are only posted once, they are cleared after successfully sending.
* A new server endpoint to track the usage data from the browser. Each 4-tuple is written to a saved object called `lens-ui-telemetry` representing the name, date, type and count of the events being tracked.
* A daily task which aggregates all the `lens-ui-telemetry` documents and saved `lens` documents into daily buckets. This task also deletes any events older than 90 days, as we are not reporting on events past that point.
* A stats collector which reads from the daily task, using the task state as a cache of data

The full list of event names that are introduced in this PR is:

Events that track a click:

```
app_filters_updated
save_failed
app_date_change
app_query_change
loaded
loaded_404
chart_switch
suggestion_confirmed
suggestion_clicked
drop_total
drop_onto_workspace
drop_onto_dimension
drop_empty
drop_non_empty
indexpattern_changed
indexpattern_filters_cleared
indexpattern_type_filter_toggled
indexpattern_existence_toggled
indexpattern_dimension_removed
indexpattern_dimension_field_changed
indexpattern_dimension_operation_<type>
indexpattern_dimension_operation_from_<type>_to_<type>
indexpattern_field_info_click
xy_change_layer_display
xy_layer_removed
xy_layer_added
```

Events that are triggered by the suggestion panel:

```
back_to_current
position_<number>_of_<number>
reload
```

Example of telemetry collected in its full structure:

```
"lens": {
  "saved_overall": {
    "bar_stacked": 1
  },
  "saved_30_days": {
    "bar_stacked": 1
  },
  "saved_90_days": {
    "bar_stacked": 1
  },
  "saved_overall_total": 1,
  "saved_30_days_total": 1,
  "saved_90_days_total": 1,
  "events_30_days": {
    "loaded": 2,
    "app_date_change": 1,
    "app_query_change": 1,
    "indexpattern_changed": 1,
    "suggestion_clicked": 1,
    "suggestion_confirmed": 1,
    "xy_layer_removed": 1
  },
  "events_90_days": {
    "loaded": 2,
    "app_date_change": 1,
    "app_query_change": 1,
    "indexpattern_changed": 1,
    "suggestion_clicked": 1,
    "suggestion_confirmed": 1,
    "xy_layer_removed": 1
  },
  "suggestion_events_30_days": {
    "position_0_of_5": 1
  },
  "suggestion_events_90_days": {
    "position_0_of_5": 1
  }
},
```

wylieconlon committed:
[Lens] Track actions in the UI by time

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3869/)
* Commit: a007ec44d25a001622bca43cfe9e6c7fcad6b8fa

wylieconlon committed:
Switch collector to use task data

chrisdavies reviewed (commented):
Just a preliminary review. It looks reasonable. My comments are mostly nits. I think we discussed all of the more relevant stuff already.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4088/)
* Commit: a832b1ec90aeaac2dc1f571e0f68c68c59dab86f

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4099/)
* Commit: 8c818e4df24977ab7434753850f83199e93af8b8

wylieconlon committed:
Report summarized version of task data when requested

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4119/)
* Commit: 5281cbdf907ff787401e77c07e79270db5a05014

wylieconlon committed:
Track a more complete set of metrics

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4139/)
* Commit: bf26faafceb87b328c492ae744e42c68424f7976

wylieconlon committed:
Collect suggestion events separately

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4173/)
* Commit: 5c7fdb717b24e07028eef99b520c78e5b6ab3ec9

wylieconlon committed:
Pre-aggregate by date in localStorage

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4191/)
* Commit: e8e2b706f28654aa27b4b0c51c031b486f3ed119

wylieconlon committed:
Add integration tests

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into lens/click-telemetry

wylieconlon committed:
Fix test linter

wylieconlon committed:
Fix telemetry naming and run at midnight instead of every minute

elasticmachine commented:
Pinging @elastic/kibana-app (Team:KibanaApp)

wylieconlon committed:
Improve cleanup at app level

wylieconlon committed:
Fix lint errors

wylieconlon committed:
Remove unused mock

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4515/)
* Commit: 863f329406b4c0db24e72a995334a8f9be164e74

wylieconlon committed:
Fix tests

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4552/)
* Commit: 7dc94b8c14c52d33d110e72ac03bb607683bd9c4

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into lens/click-telemetry

sorenlouv reviewed (approved):
apm changes lgtm

wylieconlon committed:
Fix types

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4653/)
* Commit: 6c24d16c92be700d03c6b89e7d8d3c301e11c135

wylieconlon commented:
retest

chrisdavies commented on x-pack/legacy/plugins/lens/server/usage/collectors.ts:
```diff
@@ -21,25 +21,124 @@ export function registerLensUsageCollector(
+
+function addEvents(prevEvents: Record<string, number>, newEvents: Record<string, number>) {
+  Object.keys(newEvents).forEach(key => {
+    prevEvents[key] = prevEvents[key] || 0 + newEvents[key];
```
I think this would benefit from parenthesis. It *seems* wrong to me, but I may be misunderstanding the intent. Is the intent to add `newEvents[key]` to `prevEvents[key]`? If so, I don't think this is what will happen here.

```js
// This is what is happening
prevEvents[key] = prevEvents[key] || (0 + newEvents[key]);

// I think this is what you want...
prevEvents[key] = (prevEvents[key] || 0) + newEvents[key];

```

wylieconlon replied:
This is actually a linting thing- it doesn't let me add parentheses here

chrisdavies replied:
I'm not sure why lint won't let you put parenthesis there, but I think you should, and should add a lint exception comment.

wylieconlon replied:
I must have been wrong here. I did at one point see weird issues where the linter would reformat the code, but maybe I was not doing this the right way at the time

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4689/)
* Commit: 6c24d16c92be700d03c6b89e7d8d3c301e11c135

chrisdavies commented on x-pack/legacy/plugins/lens/server/usage/collectors.ts:
```diff
@@ -21,25 +21,124 @@ export function registerLensUsageCollector(
+        const suggestionsLast30: Record<string, number> = {};
+        const suggestionsLast90: Record<string, number> = {};
+
+        const last30 = moment()
```
Your dates and suggestionDates looks identical, only for separate collections. Might be worth pulling out into a function.

chrisdavies commented on x-pack/legacy/plugins/lens/server/usage/collectors.ts:
```diff
@@ -21,25 +21,124 @@ export function registerLensUsageCollector(
-          visualization_types_overall: {},
-          visualization_types_last_30_days: {},
-          visualization_types_last_90_days: {},
+          saved_overall_total: 0,
```
I wonder if we should add some error field to the object, just so we can filter out any that are written due to an error state.

chrisdavies commented on x-pack/legacy/plugins/lens/server/usage/collectors.ts:
```diff
@@ -21,25 +21,124 @@ export function registerLensUsageCollector(
+
+  try {
+    const result = await taskManager.fetch({
+      query: { bool: { filter: { term: { _id: `task:Lens-lens_telemetry` } } } },
```
Probably should use `TASK_ID` from task.ts here instead of a hard-coded string.

chrisdavies commented on x-pack/legacy/plugins/lens/server/usage/task.ts:
```diff
@@ -0,0 +1,235 @@
+  metrics.aggregations!.daily.buckets.forEach(daily => {
+    const byType: Record<string, number> = byDateByType[daily.key] || {};
+    daily.groups.buckets.regularEvents.names.buckets.forEach(bucket => {
+      byType[bucket.key] = bucket.sums.value || 0 + (byType[daily.key] || 0);
```
It's hard for me to follow this, so am just commenting here for you to double-check that the use of `||` is really what you want here, same for below.

wylieconlon replied:
Yes, it is

chrisdavies commented on x-pack/legacy/plugins/lens/server/usage/task.ts:
```diff
@@ -0,0 +1,235 @@
+  const byDateByType: Record<string, Record<string, number>> = {};
+  const suggestionsByDate: Record<string, Record<string, number>> = {};
+
+  metrics.aggregations!.daily.buckets.forEach(daily => {
```
Are we sure this can never be undefined? Same for the deeper properties.

wylieconlon replied:
As long as the `.kibana` index exists, the `aggregations` key will be returned. And if not, this entire task will fail but that's also handled

chrisdavies commented on x-pack/legacy/plugins/lens/server/usage/task.ts:
```diff
@@ -0,0 +1,235 @@
+      query: {
+        bool: {
+          filter: [
+            { term: { type: 'lens-ui-telemetry' } },
```
I wonder if we should pull ` 'lens-ui-telemetry' ` out into a const?

chrisdavies commented on x-pack/legacy/plugins/lens/server/usage/task.ts:
```diff
@@ -0,0 +1,235 @@
+  };
+}
+
+function getNextMidnight() {
```
You're using moment everywhere else, why not here?

wylieconlon replied:
This was copied

chrisdavies commented on x-pack/legacy/plugins/lens/server/usage/types.ts:
```diff
@@ -4,11 +4,27 @@
 }
+
+export interface LensVisualizationUsage {
+  saved_overall: Record<string, number>;
```
Just to be sure this can't explode in size, we should double-check that our generated event names have a realistically small potential set of values.

wylieconlon replied:
These aren't generated names, they're all strings in code

chrisdavies reviewed (approved):
LGTM w/ tweaks.

AlonaNadler reviewed (commented):
Great work @wylieconlon I can't wait to start using it !! 
- how does the suggestions fields work e.g. ` "index_0_of_5": 1` ?
- Using ` workspace_drop_non_empty: 1` I want to know how common users drag more than 1 fields, does it make sense to do AVG(workspace_drop_non_empty/workspace_drop_empty)? 
-   `suggestion_confirmed: 1`  is it suggestion selected & saved? 
- What is the definition of `loaded`?
- What's the definition of `app_query_change` is it a change from the configuration panel? 



`

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into lens/click-telemetry

wylieconlon committed:
Update event names and fix local tracking

wylieconlon commented:
@AlonaNadler 
> how does the suggestions fields work e.g. "index_0_of_5": 1 ?

Renamed to `position_0_of_5`. All the suggestion events are also added to something like `suggestion_clicked` in the `events` list

> Using workspace_drop_non_empty: 1 I want to know how common users drag more than 1 fields, does it make sense to do AVG(workspace_drop_non_empty/workspace_drop_empty)?

No, that would not make sense. I've changed it so that there is now a number that tracks all drops regardless of their destination, and then breaks that out into:

* All drops that were empty, vs all that were non-empty
* All drops onto workspace vs all onto a dimension

Both of those should add up to 100%

> suggestion_confirmed: 1 is it suggestion selected & saved?

This is number of clicks on the "Reload" button

> What is the definition of loaded?

This tracks any page load, including ones that happen because the user saved. I can change this definition.

> What's the definition of app_query_change is it a change from the configuration panel?

For example, a KQL filter was typed into the query bar.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4786/)
* Commit: fd1f6d6577a703b9f0673551846af658b606e981

chrisdavies commented on x-pack/legacy/plugins/lens/public/drag_drop/drag_drop.tsx:
```diff
@@ -121,7 +122,8 @@ export function DragDrop(props: Props) {
 
-    if (onDrop) {
+    if (onDrop && droppable) {
+      trackUiEvent('drop_total');
```
This is probably fine, but if we ever want to make this drag / drop thing more general, this would be something that requires yanking out.

chrisdavies commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/dimension_panel/dimension_panel.tsx:
```diff
@@ -137,6 +138,10 @@ export const IndexPatternDimensionPanel = memo(function IndexPatternDimensionPan
 
+          trackUiEvent('drop_onto_dimension');
+          const hasData = Object.values(props.state.layers).some(({ columns }) => columns.length);
+          trackUiEvent(hasData ? 'drop_non_empty' : 'drop_empty');
```
This is fine, but also, do we need three different events? drop_onto_dimension would be a sum of drop_non_empty and drop_empty

wylieconlon replied:
It's actually five events:

`drop_total` is every drop

Two pairs:

`drop_onto_workspace` + `drop_onto_dimension` = `drop_total`
`drop_empty` + `drop_non_empty` = `drop_total`

chrisdavies commented on x-pack/legacy/plugins/lens/public/lens_ui_telemetry/factory.ts:
```diff
@@ -0,0 +1,134 @@
+  }
+
+  private async postToServer() {
+    this.readFromStorage();
```
This is probably fine, but it does introduce a disk read for a whole bunch of user-driven events. I wonder if this'll have any impact on performance on low-end devices... Probably not.

Anyway, an alternative would be to track all the stuff in memory, but when you flush it to storage, read and merge then, and reset your in-memory state.

wylieconlon replied:
There are probably some alternatives here, but this is the simplest implementation of it

chrisdavies reviewed (approved):
LGTM w/ tweaks. I think you need to correct the operator precedence stuff I commented on in this review and my previous one. But other than that, I think this is fine.

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into lens/click-telemetry

wylieconlon committed:
Respond to review comments

wylieconlon committed:
Fix task time

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into lens/click-telemetry

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4890/)
* Commit: 16ebf6c36a13807683969827a3b6b8fbd41225e4

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4908/)
* Commit: 4da50b3957bdd53e02c15e64926cd3c930615d22

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4919/)
* Commit: b5d54a769651e51e8cff0ad7de56f19f9e9a0f0e

wylieconlon committed:
Fix test

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into lens/click-telemetry

Bamieh reviewed (commented):
This collection method almost duplicates @kbn/analytics code just for providing a new collection method. This UI metric can be easily baked into the `@kbn/analytics` library by introducing it as a new `METRIC_TYPE`. Doing so it makes it easier to manage telemetry, refactor code and have others utilize these metric types.

I have opened an issue for this as a follow up to this PR: https://github.com/elastic/kibana/issues/48352

While it is nice to have this tracking as soon as possible for lens. Ideally implementing this would be approached differently; I've opened an issue so we can revisit the use of this metric type so we can learn its usefulness for the consumers of this usage data and then introduce it as `METRIC_TYPE` in `@kbn/analytics` for others to be able to use it as well.

I have opened an issue for this as a follow up to this PR: https://github.com/elastic/kibana/issues/48353

While not completely thrilled about introducing new telemetry code floating outside our telemetry packages (ui_metric / telemetry / and `@kbn/analytics`); I believe we can merge this and revisit these concerns afterwards.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4933/)
* Commit: 1836ed1c93d56efb146aab6531383a9b07b90c6d

wylieconlon commented:
@Bamieh You are right that it is heavily based on the existing code, and I am looking forward to integrating some of these changes for the rest of Kibana to use in releases.
